rfcs/2021-09-01-8547-accept-metrics-in-datadog-agent-source.md

```diff
@@ -0,0 +1,168 @@
+- `/api/v1/series` & `/api/beta/sketches` for metrics submission
+
+
+
+### Implementation
+
+A few details about the Datadog Agents & [Datadog metrics](https://docs.datadoghq.com/metrics/types/):
+- The base structure for all metrics is named
+  [`MetricSample`](https://github.com/DataDog/datadog-agent/blob/main/pkg/metrics/metric_sample.go#L81-L94) and can be
+  of [several types](https://github.com/DataDog/datadog-agent/blob/main/pkg/metrics/metric_sample.go#L20-L31)
+- Major Agent usecases:
+  - Metrics are send from corechecks (i.e. go code)
+    [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/aggregator/sender.go#L227-L252)
+  - Dogstatsd metrics are converted to the `MetricSample` structure
+    [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/dogstatsd/enrich.go#L87-L137) However Datadog Agents
+    metrics are transformed before being sent, ultimately metrics accounts for two differents kind of payload:
+- The count, gauge and rate series kind of payload, sent to `/api/v1/series` using the [JSON schema officially
+  documented](https://docs.datadoghq.com/api/latest/metrics) with few undocumented [additional
+  fields](https://github.com/DataDog/datadog-agent/blob/main/pkg/metrics/series.go#L45-L57), but this align very well
+  with the existing `datadog_metrics` sinks.
+- The sketches kind of payload, sent to `/api/beta/sketches` and serialized as protobuf as shown
+  [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/serializer/serializer.go#L315-L338) (it ultimately lands
+  [here](https://github.com/DataDog/datadog-agent/blob/main/pkg/metrics/sketch_series.go#L103-L269)). Public `.proto`
+  definition can be found
+  [here](https://github.com/DataDog/agent-payload/blob/master/proto/metrics/agent_payload.proto#L47-L81).
+
+Vector has a nice description of its [metrics data
+model](https://vector.dev/docs/about/under-the-hood/architecture/data-model/metric/) and a [concise enum for
+representing it](https://github.com/timberio/vector/blob/master/lib/vector-core/src/event/metric.rs#L135-L169).
+
+
+The implementation would the consist in:
+
+- Implement in the `datadog_agent` source the ability to proxy all known route (see above) as-is directly to Datadog
+  (this will require an new settings in the Vector source, something like `unused_payload_site: us3.datadoghq.com`). At
+  the beginning it will also proxify `/api/v1/series` & `/api/beta/sketches`.
+- Handle the `/api/v1/series` route (based on both the offical [API](https://docs.datadoghq.com/api/latest/metrics/) and
+  the [Datadog Agent itself](https://github.com/DataDog/datadog-agent/blob/main/pkg/forwarder/telemetry.go#L20-L31)) to
+  cover every metric type handled by this endpoint (count, gauge and rate) and:
+    - Add support for missing fields in the `datadog_metrics` sinks
+    - Align tag manipulation (Datadog allows `key:foo` & `key:bar`, and the tags are encoded as a list whereas Vector
+      encode those as a map).
```

Comment:
I know that the way DD handles re-used tags is not documented and thus might differ for different places(logs, metrics, traces etc). 
Tags with no values  are supported, but it is not generally recommended. From [here](https://docs.datadoghq.com/getting_started/tagging/#defining-tags): "For optimal functionality, Datadog recommends constructing tags in the <KEY>:<VALUE> format." 
Though I don't have any statistics(and probably it's not available) but I believe these two cases are quite uncommon.

Reply:
I don't think it's very common to have multiple key/value tag with the same key (and there are some piece, e.g. jmxfetch that dont support that because they internally use a [string/string map](https://github.com/DataDog/jmxfetch/blob/master/src/main/java/org/datadog/jmxfetch/Instance.java#L91) to store tags. I think putting a note about that particular case not being supported is okay. To keep a map structure to store tags we could utlimately adjust the map to support a list of string for the value and specify the behaviour on a per sink/source basis if we have 0/1/2+ items in the list for a given key.

I have no idea about no-value tags, but I don't think it is an issue to store an empty string for the value from a Vector perspective ?

Reply:
I would say that we can probably limit ourselves to matching Datadog behavior for recommended use cases, at least at first, which seems like it wouldn't require much change. But I'm definitely open to adjusting how we store tags so that things like this can just be passed through to the API where possible. Sinks like prom would need to resolve any differences in data model themselves in that case. I'm not sure how large a project that would be, but I would assume not massive.

Reply:
I'm with @lukesteensen on this. If the *recommended* use cases match our internal representation with minimal modification, then limiting our support to that subset is the best plan until there is further demand. If we need to support multi-value keys, we can consider alternate representations.

Reply:
I adjusted the RFC to reflect this. We will see if there is demand for use case like that later.